File: python/src/lino_rest_api/cors.py
"""
Cross-origin resource sharing (specification section 8).
"""

from typing import Any

#: Headers a browser must be allowed to send to negotiate Links Notation.
DEFAULT_ALLOWED_HEADERS = [
    "Content-Type",
    "Accept",
    "Authorization",
    "If-Match",
    "If-None-Match",
]

#: Headers a browser must be allowed to read to follow the protocol.
DEFAULT_EXPOSED_HEADERS = ["ETag", "Link", "Location", "Allow"]

DEFAULT_METHODS = ["GET", "HEAD", "OPTIONS", "POST", "PUT", "PATCH", "DELETE"]


def cors_headers(
    options: dict[str, Any] | bool | None = None,
    request_origin: str | None = None,
) -> dict[str, str]:
    """
    Build the CORS response headers for a request.

    Args:
        options: Policy with ``origin``, ``methods``, ``allowed_headers``,
            ``exposed_headers``, ``credentials`` and ``max_age`` members; True
            selects the permissive default policy
        request_origin: ``Origin`` header of the request

    Returns:
        Response headers ({} when the origin is not allowed)
    """
    policy: dict[str, Any] = {} if options in (None, True, False) else dict(options)

    origin = policy.get("origin", "*")
    methods = policy.get("methods", DEFAULT_METHODS)
    allowed_headers = policy.get("allowed_headers", DEFAULT_ALLOWED_HEADERS)
    exposed_headers = policy.get("exposed_headers", DEFAULT_EXPOSED_HEADERS)
    credentials = policy.get("credentials", False)
    max_age = policy.get("max_age", 600)

    allowed_origins = origin if isinstance(origin, list | tuple) else [origin]
    allow_origin = None
    if "*" in allowed_origins:
        allow_origin = request_origin if credentials and request_origin else "*"
    elif request_origin and request_origin in allowed_origins:
        allow_origin = request_origin

    if not allow_origin:
        return {}

    headers = {
        "Access-Control-Allow-Origin": allow_origin,
        "Access-Control-Allow-Methods": ", ".join(methods),
        "Access-Control-Allow-Headers": ", ".join(allowed_headers),
        "Access-Control-Expose-Headers": ", ".join(exposed_headers),
        "Access-Control-Max-Age": str(max_age),
    }
    if credentials:
        headers["Access-Control-Allow-Credentials"] = "true"
    return headers

File: rust/tests/cors.rs
//! Tests for cross-origin requests (specification section 8), ported from
//! `python/tests/test_cors.py`.

use lino_rest_api::cors::{
    CorsPolicy, DEFAULT_ALLOWED_HEADERS, DEFAULT_EXPOSED_HEADERS, cors_headers,
};
use lino_rest_api::headers::Headers;
use lino_rest_api::middleware::append_vary;

#[test]
fn content_type_and_accept_are_allowed_so_browsers_can_negotiate_lino() {
    assert!(DEFAULT_ALLOWED_HEADERS.contains(&"Content-Type"));
    assert!(DEFAULT_ALLOWED_HEADERS.contains(&"Accept"));
    assert!(DEFAULT_ALLOWED_HEADERS.contains(&"If-Match"));
}

#[test]
fn etag_and_link_are_exposed_so_clients_can_follow_the_protocol() {
    assert!(DEFAULT_EXPOSED_HEADERS.contains(&"ETag"));
    assert!(DEFAULT_EXPOSED_HEADERS.contains(&"Link"));
}

#[test]
fn the_default_policy_allows_any_origin() {
    let headers = cors_headers(&CorsPolicy::default(), Some("https://example.com"));
    assert_eq!(headers.get("Access-Control-Allow-Origin"), Some("*"));
    assert!(
        headers
            .get("Access-Control-Allow-Methods")
            .unwrap()
            .contains("PATCH")
    );
}

#[test]
fn an_allow_list_reflects_the_request_origin() {
    let policy = CorsPolicy::permissive().with_origins(["https://example.com"]);
    let allowed = cors_headers(&policy, Some("https://example.com"));
    assert_eq!(
        allowed.get("Access-Control-Allow-Origin"),
        Some("https://example.com")
    );
    assert!(cors_headers(&policy, Some("https://evil.example")).is_empty());
}

#[test]
fn credentialed_requests_echo_the_origin() {
    let policy = CorsPolicy::permissive().with_credentials(true);
    let headers = cors_headers(&policy, Some("https://example.com"));
    assert_eq!(
        headers.get("Access-Control-Allow-Origin"),
        Some("https://example.com")
    );
    assert_eq!(
        headers.get("Access-Control-Allow-Credentials"),
        Some("true")
    );
}

#[test]
fn append_vary_keeps_existing_field_names() {
    let mut headers = Headers::new();
    append_vary(&mut headers, "Accept");
    append_vary(&mut headers, "Origin");
    append_vary(&mut headers, "accept");
    assert_eq!(headers.get("Vary"), Some("Accept, Origin"));
}
